### PR TITLE
[Feature branch] Datasets rest and sqlalchemy backend

### PR DESCRIPTION
**Pull request by prithvikannan**
## Related Issues/PRs

https://github.com/mlflow/mlflow/pull/8146

## What changes are proposed in this pull request?

Implement `log_inputs` REST and Python SQLAlchemy backend. Update `get_run` and `search_runs` to return run inputs. 

## How is this patch tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [x] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [x] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by prithvikannan** [comment]
@mlflow-automation autoformat

File: mlflow/store/tracking/dbmodels/models.py [hunk 1]
```diff
@@ -537,10 +538,10 @@ def to_mlflow_entity(self):
...
             name=self.name,
             digest=self.digest,
-            source_type=self.source_type,
-            source=self.source,
-            schema=self.schema,
-            profile=self.profile,
+            source_type=self.dataset_source_type,
+            source=self.dataset_source,
+            schema=self.dataset_schema,
+            profile=self.dataset_profile,
```
Comment (prithvikannan) [inline mlflow/store/tracking/dbmodels/models.py:544]: these were misnamed earlier. caught this when writing the `log_inputs` test

File: mlflow/store/tracking/dbmodels/models.py [hunk 2]
```diff
@@ -513,7 +514,7 @@ class SqlDataset(Base):
     """
     dataset_profile = Column(UnicodeText, nullable=True)
     """
-    Param dataset_schema: `UnicodeText`.
+    Param dataset_profile: `UnicodeText`.
```
Comment (prithvikannan) [inline mlflow/store/tracking/dbmodels/models.py:517]: fixing typo from earlier

File: mlflow/store/tracking/sqlalchemy_store.py [hunk 3]
```diff
@@ -515,6 +521,49 @@ def _get_run(self, session, run_uuid, eager=False):
 
         return runs[0]
 
+    def _get_run_inputs(self, session, run_uuids):
+        SqlDatasetAlias = sqlalchemy.orm.aliased(SqlDataset)
+        datasets = (
+            session.query(
+                SqlInput.input_uuid, SqlInput.destination_id.label("run_uuid"), SqlDataset
+            )
+            .join(SqlDatasetAlias, SqlInput.source_id == SqlDatasetAlias.dataset_uuid)
```
Comment (dbczumar) [inline mlflow/store/tracking/sqlalchemy_store.py:530]: Nit: Can we also include `source_type == DATASET` in the query?

File: mlflow/store/tracking/sqlalchemy_store.py [hunk 4]
```diff
@@ -515,6 +521,49 @@ def _get_run(self, session, run_uuid, eager=False):
 
         return runs[0]
 
+    def _get_run_inputs(self, session, run_uuids):
+        SqlDatasetAlias = sqlalchemy.orm.aliased(SqlDataset)
+        datasets = (
```
Comment (dbczumar) [inline mlflow/store/tracking/sqlalchemy_store.py:526]: Nit:

```suggestion
        dataset_inputs = (
```

File: mlflow/store/tracking/sqlalchemy_store.py [hunk 5]
```diff
@@ -515,6 +521,49 @@ def _get_run(self, session, run_uuid, eager=False):
...
+        input_uuids = [str(dataset.input_uuid) for dataset in datasets]
+        SqlInputTagAlias = sqlalchemy.orm.aliased(SqlInputTag)
+        input_tags = (
+            session.query(
+                SqlInput.input_uuid, SqlInput.destination_id.label("run_uuid"), SqlInputTag
+            )
+            .join(SqlInputTagAlias, (SqlInput.input_uuid == SqlInputTagAlias.input_uuid))
+            .filter(SqlInput.input_uuid.in_(input_uuids))
+            .order_by("run_uuid")
+        ).all()
```
[the remark this answers is not in the document]
Comment (prithvikannan) [reply mlflow/store/tracking/sqlalchemy_store.py:543]: Per offline convo, we don't want to do this. Parsing this response will become very complex

File: mlflow/store/tracking/sqlalchemy_store.py (hunk 4 above)
Comment (prithvikannan) [reply mlflow/store/tracking/sqlalchemy_store.py:526]: This should still be datasets. We are only fetching datasets in this call, and form the `dataset_inputs` once we have the input tags.

Comment (dbczumar) [reply mlflow/store/tracking/sqlalchemy_store.py:526]: You're right, whoops :)

File: mlflow/store/tracking/dbmodels/models.py [hunk 6]
```diff
@@ -598,7 +598,7 @@ class SqlInputTag(Base):
     __tablename__ = "input_tags"
     __table_args__ = (PrimaryKeyConstraint("input_uuid", "name", name="input_tags_pk"),)
 
-    input_uuid = Column(String(36), ForeignKey("experiments.experiment_id"), nullable=False)
+    input_uuid = Column(String(36), ForeignKey("inputs.input_uuid"), nullable=False)
```
Comment (prithvikannan) [inline mlflow/store/tracking/dbmodels/models.py:601]: also a typo from earlier

Comment (dbczumar) [reply mlflow/store/tracking/dbmodels/models.py:601]: Any migration typos, or just in the model?

**Review by dbczumar — APPROVED** [review]
LGTM! Thanks @prithvikannan !

**Pull request merged**